### DISCUSSION TITLE
make contributions of translations easier

### DISCUSSION THREAD

**Discussion by samuelstroschein** [Laravel-Lang/common#12] [Q&A]
Hi everyone, 

You built a lot of tooling like auto-translations, etc. yourself. Inspiring to see as the project I maintain, [inlang/inlang]( https://github.com/inlang/inlang), strives to bring such tooling to everyone.  

I am wondering whether the current state of inlang could already be of use for laravel. Contributors seem to be required to manually edit files or discussion are used to contribute translations https://github.com/Laravel-Lang/lang/discussions/1421. I could write a plugin that provides contributors with a UI for the translations in this repository, see  https://inlang.com/editor/github.com/inlang/example for example. 

There would be no overhead of maintainers or contributors as inlang is built on git. No accounts (login with GitHub), no sync, no pipelines are required. Contributions open a PR. More tools like the one you built in this repo will be part of inlang soon. 

Would an UI over the translations in this repo be of benefit to the laravel community?

**Comment by andrey-helldar** [comment]
On the one hand, the mechanics of inlang is interesting and useful, but on the other hand, we will face many problems starting with the number of supported localizations - at the moment we have 78 of them.

Also, the other day we implemented a machine translation mechanism using Google Translate for untranslated keys. After that, anyone can easily translate any of the languages.

Another problem that an inlang project can face is line volume. The Laravel Lang project can support multiple versions of each package at the same time, so the final file has a fairly large volume of keys. For example [locales/af/json.json](https://github.com/Laravel-Lang/lang/blob/main/locales/af/json.json). This file currently contains 1100 translation keys.

In addition, if a referent is assigned to one of the localizations, then the review code is assigned to it, because this person knows the language and can determine the correct application of grammar. When a person proposes a large enough amount of corrections, we can invite him to join our team. In the case of the inlang project, this behavior can be corrected, but how then will key translations come to the Laravel Lang project to pass the code review?

**Reply by samuelstroschein** [reply]
Andrey, thanks for the response. 

None of your existing workflows would need to change. 

> [...] how then will key translations come to the Laravel Lang project to pass the code review?

Translators directly commit/open PRs to the repo, just as if they would be developers/manually editing files. 

Your CI/CD (machine translations) will also work hand-in-hand with your code review workflow. As a matter of fact, that is exactly why I believe inlang makes a difference. Everything works on top of git which entails unprecedented collaboration and automation. For example, translators (unbeknown to them or not) create commits, which can be (if the translator is a direct member of the repo) pushed to a specific branch, or a pull request can be opened for review. Just like software engineers would contribute to a repository. [The next git](https://inlang.com/documentation/the-next-git) might be of interest in that context.  

> we will face many problems starting with the number of supported localizations - at the moment we have 78 of them 

Rest assured, (the performance implications) of supporting many localization should be our problem :D 

That said, we do have performance issues with larger repositories right now. This repository is 30MB large and should be fine. The main issue is the underlying git implementation that we use. A project to switch to [libgit2](https://libgit2.org/) is on its way, see https://github.com/orgs/inlang/projects/9. 



Looking forward to your reply of whether inlang would be helpful and if not why not/what you are missing

**Reply by andrey-helldar** [reply]
In the current state, the total weight of localization files is 9.24 Mb.

The only thing that confuses me is the translated volume, taking into account 78 supported localizations. In the case of opening a file in the IDE, all lines go one after another within the same localization. The site also offers to display keys for everyone with the ability to select the ones you need.

However, I am interested in the development of the project.

@Laravel-Lang/laravel-lang, guys, what do you think about this discussion?

**Reply by samuelstroschein** [reply]
Oh, I assume that you refer to the UI of listing all Messages beneath each other. Totally agree. 

It's "the first version" that offers filters in the top right but has a long way to go. For example, the Messages are not broken into their components (text vs placeholders) yet. The latter would massively decrease the risk of broken translations. Auto-generating screenshots are on the list as well. A lot to build!  If you know UI/UX designers or engineers that are interested in improving localization to the point that localization becomes so easy that the web turns multilingual, please let me know.

**Reply by andrey-helldar** [reply]
I will ask, but I do not promise that they will agree.

**Reply by caouecs** [reply]
the project is interesting but with the number of strings ;) we need to test I think

**Reply by samuelstroschein** [reply]
@caouecs I will open a test PR to see whether we are running into performance issues later today. Fingers crossed 🤞

**Reply by samuelstroschein** [reply]
Will do the PR tomorrow. Didn't get to it today

**Reply by samuelstroschein** [reply]
Alright, it's 4 am in Berlin. Finally got to test inlang with this repo. 

Crossing our fingers did not help. The number of languages/strings crash Chrome. The cause is unknown (error 5 in Chrome). Choosing a subset of the languages works, see https://inlang.com/editor/github.com/samuelstroschein/lang. But, even if Chrome would not crash, the user experience would be ... improvable due to performance issues: 

- Cloning the repo into the browser takes 3 minutes. 
- Every commit will lead to a few seconds of loading time 

The performance issues are aggregated under [project #9](https://github.com/orgs/inlang/projects/9).

While we are busy fixing the performance issues, which can take months as we are hacking git to be a backend (see [the next git](https://inlang.com/documentation/the-next-git)), how could inlang help laravel otherwise/feedback/ideas on what would make your and your communities easier with regards to localization?
